File: src/test/java/cz/vhromada/catalog/service/ProgramServiceTest.java
package cz.vhromada.catalog.service;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import cz.vhromada.catalog.domain.Program;
import cz.vhromada.catalog.repository.ProgramRepository;
import cz.vhromada.catalog.utils.ProgramUtils;
import cz.vhromada.common.repository.MovableRepository;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.service.MovableServiceTest;

import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.cache.Cache;

/**
 * A class represents test for class {@link ProgramService}.
 *
 * @author Vladimir Hromada
 */
class ProgramServiceTest extends MovableServiceTest<Program> {

    /**
     * Instance of {@link ProgramRepository}
     */
    @Mock
    private ProgramRepository repository;

    /**
     * Test method for {@link ProgramService#ProgramService(ProgramRepository, Cache)} with null repository for programs.
     */
    @Test
    void constructor_NullProgramRepository() {
        assertThatThrownBy(() -> new ProgramService(null, getCache())).isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link ProgramService#ProgramService(ProgramRepository, Cache)} with null cache.
     */
    @Test
    void constructor_NullCache() {
        assertThatThrownBy(() -> new ProgramService(repository, null)).isInstanceOf(IllegalArgumentException.class);
    }

    @Override
    protected MovableRepository<Program> getRepository() {
        return repository;
    }

    @Override
    protected MovableService<Program> getService() {
        return new ProgramService(repository, getCache());
    }

    @Override
    protected String getCacheKey() {
        return "programs";
    }

    @Override
    protected Program getItem1() {
        return ProgramUtils.newProgramDomain(1);
    }

    @Override
    protected Program getItem2() {
        return ProgramUtils.newProgramDomain(2);
    }

    @Override
    protected Program getAddItem() {
        return ProgramUtils.newProgramDomain(null);
    }

    @Override
    protected Program getCopyItem() {
        final Program program = ProgramUtils.newProgramDomain(null);
        program.setPosition(0);

        return program;
    }

    @Override
    protected Class<Program> getItemClass() {
        return Program.class;
    }

    @Override
    protected void assertDataDeepEquals(final Program expected, final Program actual) {
        ProgramUtils.assertProgramDeepEquals(expected, actual);
    }

}

File: src/test/kotlin/cz/vhromada/catalog/service/ProgramServiceTest.kt
package cz.vhromada.catalog.service

import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.verifyNoMoreInteractions
import com.nhaarman.mockitokotlin2.whenever
import cz.vhromada.catalog.domain.Program
import cz.vhromada.catalog.repository.ProgramRepository
import cz.vhromada.catalog.utils.ProgramUtils
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.test.service.MovableServiceTest
import cz.vhromada.common.test.utils.TestConstants
import org.mockito.Mock
import org.springframework.data.jpa.repository.JpaRepository

/**
 * A class represents test for class [ProgramService].
 *
 * @author Vladimir Hromada
 */
class ProgramServiceTest : MovableServiceTest<Program>() {

    /**
     * Instance of [ProgramRepository]
     */
    @Mock
    private lateinit var repository: ProgramRepository

    /**
     * Instance of [AccountProvider]
     */
    @Mock
    private lateinit var accountProvider: AccountProvider

    /**
     * Instance of [TimeProvider]
     */
    @Mock
    private lateinit var timeProvider: TimeProvider

    override fun getRepository(): JpaRepository<Program, Int> {
        return repository
    }

    override fun getAccountProvider(): AccountProvider {
        return accountProvider
    }

    override fun getTimeProvider(): TimeProvider {
        return timeProvider
    }

    override fun getService(): MovableService<Program> {
        return ProgramService(repository, accountProvider, timeProvider, cache)
    }

    override fun getCacheKey(): String {
        return "programs${TestConstants.ACCOUNT.id}"
    }

    override fun getItem1(): Program {
        return ProgramUtils.newProgramDomain(1)
    }

    override fun getItem2(): Program {
        return ProgramUtils.newProgramDomain(2)
    }

    override fun getAddItem(): Program {
        return ProgramUtils.newProgramDomain(null)
    }

    override fun getCopyItem(): Program {
        return ProgramUtils.newProgramDomain(null)
                .copy(position = 0)
    }

    override fun initAllDataMock(data: List<Program>) {
        whenever(repository.findByAuditCreatedUser(any())).thenReturn(data)
    }

    override fun verifyAllDataMock() {
        verify(repository).findByAuditCreatedUser(TestConstants.ACCOUNT_ID)
        verifyNoMoreInteractions(repository)
    }

    override fun anyItem(): Program {
        return any()
    }

    override fun argumentCaptorItem(): KArgumentCaptor<Program> {
        return argumentCaptor()
    }

    override fun assertDataDeepEquals(expected: Program, actual: Program) {
        ProgramUtils.assertProgramDeepEquals(expected, actual)
    }

}
